Check: version grep: filter out matches on store path

martinetd opened the pull request:
Use a negative lookahead on store path to avoid false positive on version grep when there are wrappers that contain the full store path and that store path includes the version.

Link: NixOS/nixpkgs#186698

-----

Note I've never done any haskell and there most probably are more natural ways of doing it, but it should work as far as I can tell.
Happy to adjust anything :)

ju1m commented on src/Check.hs:
```diff
@@ -90,13 +91,30 @@ checkTestsBuildReport False =
   "- The tests defined in `passthru.tests`, if any, passed"
 
+versionWithoutPath :: String -> Text -> String
+versionWithoutPath resultPath expectedVersion = do
```
`do` (syntactic sugar to insert the `Monad` operators `>>` or `>>=`) is useless here, you're returning a `String` not a `Monad m => m String`

ju1m commented on src/Check.hs:
```diff
@@ -90,13 +91,30 @@ checkTestsBuildReport False =
+  -- /^(?!.*${storePathWithoutVersion}).*${version}/
+  -- Note we also replace '.' with '\\.' for literal match in grep -P,
+  -- there might be a better way of escaping everything...
+  let storePath = maybe (T.pack resultPath) id $ T.stripPrefix (T.pack "/nix/store/") (T.pack resultPath)
```
`T.pack` on string literals is useless here because the language extension `OverloadedStrings` is enabled on this file (see the top of the file).

ju1m replied:
`maybe foo id` is more idiomatically `fromMaybe foo`

ju1m commented on src/Check.hs:
```diff
@@ -90,13 +91,30 @@ checkTestsBuildReport False =
+  "^(?!.*"
+    <> T.unpack storePathWithoutVersion
+    <> ").*"
+    <> T.unpack (T.replace (T.pack ".") (T.pack "\\.") expectedVersion)
```
Side note: appending (`<>`) multiple times is often less efficient than concatenating `T.concat [foo, bar, baz]`, but it would be over optimizing here. Also, for that kind of generation, one would usually leverage an interpolation library, like [PyF](https://hackage.haskell.org/package/PyF). Here `nixpkgs-update` uses in a few places [https://hackage.haskell.org/package/neat-interpolation](neat-interpolation), but it produces only `Text` and I was unable to do that (it was always complaining that some variables were not in scope, despite them actually being in scope), at least with GHC 9.0.2.

rhendric commented:
Is `^(?!.*${storePathWithoutVersion}).*${version}` truly what you mean to do here? Or would `(?<!${storePathWithoutVersion})${version}` suffice? (Imagine the substring `libpath=/nix/store/z9l2xakz7cgw6yfh83nh542pvc0g4rkq-geeqie-2.0.1/lib;version=2.0.1` exists somewhere in the package; should this be a positive or negative result?)

Also, since you're breaking out the PCRE engine for this, the better way to escape everything is to surround literals with `\Q` and `\E` as long as those literals don't contain any backslashes.

martinetd commented:
I've also adjusted the function to look for version a bit further than just suffix, and return the version string only if no match. That'll help with `hash-foo-1.2.3-dev` style paths.

martinetd commented:
Hello -- just doing a round-up of my old opened PR.

I think I've addressed all the concerns, please tell me if there's anything I can do to help merge this. I've had a rough time due to false positives on the update script, would be great to be using this instead :)

ryantm reviewed (approved):
thanks!
